diff --git a/apps/api/controllers/authController.js b/apps/api/controllers/authController.js
index b7b02b4..b7c0eba 100644
--- a/apps/api/controllers/authController.js
+++ b/apps/api/controllers/authController.js
@@ -1,38 +1,67 @@
-import { refreshService, registerService } from '../services/authService.js';
-import { passport, issueTokens } from '../utils/passport.js';
+import { passport } from "../utils/passport.js";
+import { logoutService, refreshService, registerService, signTokensService } from "../services/authService.js";
 
-export const register = async (req, res) => {
+export const issueTokens = async (req, res) => {
+  const { accessToken, refreshToken } = req.locals;
+  if (!accessToken || !refreshToken) {
+    return res.status(500).json({ message: "Tokens not prepared" });
+  }
+
+  res.cookie("refreshToken", refreshToken, {
+    httpOnly: true,
+    secure: process.env.NODE_ENV === "production",
+    sameSite: "strict",
+    path: "/auth/refresh",
+    maxAge: 7 * 24 * 60 * 60 * 1000,
+  });
+
+  return res.json({ accessToken });
+};
+
+export const register = async (req, res, next) => {
   try {
     const { email, password, fullName } = req.body;
     const user = await registerService(email, password, fullName);
-    return issueTokens(res, user.id);
+    Object.assign(res.locals, { user, ...signTokensService(user) });
+    next();
   } catch (error) {
-    return res
-      .status(500)
-      .json({ message: error.message || 'Registration failed' });
+    return res.status(400).json({ message: error.message || "Registration failed" });
   }
 };
 
-export const login = (req, res, next) => {
-  return passport.authenticate(
-    'local',
-    { session: false },
-    (err, user, info) => {
-      if (err || !user) {
-        return res.status(400).json(info || { message: 'Login failed' });
-      }
-      return issueTokens(res, user.id);
+export const localLogin = (req, res, next) => {
+  return passport.authenticate("local", { session: false }, (err, user, info) => {
+    if (err || !user) {
+      return res.status(400).json(info || { message: "Login failed" });
     }
-  )(req, res, next);
+    Object.assign(res.locals, { user, ...signTokensService(user) });
+    next();
+  })(req, res, next);
+};
+
+export const oauthLogin = async (req, res, next) => {
+  Object.assign(res.locals, { user: req.user, ...signTokensService(req.user) });
+  next();
+};
+
+export const refresh = async (req, res, next) => {
+  try {
+    const oldRefreshToken = req.cookies?.refreshToken;
+    Object.assign(res.locals, { ...(await refreshService(oldRefreshToken)) });
+    next();
+  } catch (err) {
+    console.error(err);
+    return res.status(401).json({ message: "Not logged in" });
+  }
 };
 
-export const refresh = async (req, res) => {
+export const logout = async (req, res) => {
   try {
     const refreshToken = req.cookies?.refreshToken;
-    const user = await refreshService(refreshToken);
-    return issueTokens(res, user.id);
+    await logoutService(refreshToken);
+    return res.json({ message: "Logout successfully" });
   } catch (err) {
     console.error(err);
-    return res.status(401).json({ message: 'Not logged in' });
+    return res.status(500).json({ message: err.message || "Logout failed" });
   }
 };
diff --git a/apps/api/package.json b/apps/api/package.json
index 8306c48..90452b7 100644
--- a/apps/api/package.json
+++ b/apps/api/package.json
@@ -14,7 +14,12 @@
     "passport-google-oauth20": "^2.0.0",
     "passport-jwt": "^4.0.1",
     "passport-local": "^1.0.0",
-    "zod": "^4.1.11"
+    "zod": "^4.1.11",
+    "bcryptjs": "^3.0.2",
+    "cors": "^2.8.5",
+    "jsonwebtoken": "^9.0.2",
+    "node-cron": "^4.2.1",
+    "nodemailer": "^7.0.6"
   },
   "scripts": {
     "dev": "node server.js",
@@ -31,4 +36,4 @@
     "nodemon": "^3.1.10",
     "prisma": "^6.16.2"
   }
-}
+}
\ No newline at end of file
diff --git a/apps/api/prisma/schema.prisma b/apps/api/prisma/schema.prisma
index b4a29e2..163705d 100644
--- a/apps/api/prisma/schema.prisma
+++ b/apps/api/prisma/schema.prisma
@@ -319,4 +319,9 @@ model DiscountCode {
   
   createdAt   DateTime      @default(now())
   updatedAt   DateTime      @updatedAt
+}
+
+model RevokedToken {
+  jti String @id @default(auto()) @map("_id") @db.ObjectId
+  revokedAt DateTime @default(now())
 }
\ No newline at end of file
diff --git a/apps/api/routes/authRoutes.js b/apps/api/routes/authRoutes.js
index dabfe2f..549871e 100644
--- a/apps/api/routes/authRoutes.js
+++ b/apps/api/routes/authRoutes.js
@@ -1,24 +1,20 @@
-import express from "express";
-import { login, refresh, register } from "../controllers/authController.js";
-import { passport, issueTokens } from "../utils/passport.js";
-import { validate } from "../middleware/zodMiddleware.js";
+import { Router } from "express";
+import { passport } from "../utils/passport.js";
 import { registerSchema } from "../schemas/userSchema.js";
+import { validate } from "../middleware/zodMiddleware.js";
+import { issueTokens, localLogin, oauthLogin, refresh, register } from "../controllers/authController.js";
 
-const authRouter = express.Router();
+const authRouter = Router();
 
-authRouter.post("/refresh", refresh);
+authRouter.post("/refresh", refresh, issueTokens);
 
 authRouter.post("/register", validate(registerSchema), register);
-authRouter.post("/login", login);
+authRouter.post("/login", localLogin, issueTokens);
 
 authRouter.get("/google", passport.authenticate("google", { scope: ["profile", "email"] }));
-authRouter.get("/google/callback", passport.authenticate("google", { session: false }), (req, res) => {
-  return issueTokens(res, req.user.id);
-});
+authRouter.get("/google/callback", passport.authenticate("google", { session: false }), oauthLogin, issueTokens);
 
 authRouter.get("/facebook", passport.authenticate("facebook", { scope: ["email"] }));
-authRouter.get("/facebook/callback", passport.authenticate("facebook", { session: false }), (req, res) => {
-  return issueTokens(res, req.user.id);
-});
+authRouter.get("/facebook/callback", passport.authenticate("facebook", { session: false }), oauthLogin, issueTokens);
 
 export default authRouter;
diff --git a/apps/api/routes/brandRoutes.js b/apps/api/routes/brandRoutes.js
index 04c1af5..21d6cd8 100644
--- a/apps/api/routes/brandRoutes.js
+++ b/apps/api/routes/brandRoutes.js
@@ -2,17 +2,18 @@ import express from "express";
 import { passport } from "../utils/passport.js";
 
 import checkAdmin from "../middleware/adminMiddleware.js";
+import { requireAuth } from "../middleware/authMiddleware.js";
 import { createBrand, deleteBrand, getBrandBySlug, getBrands, getBrandsFromTrash, restoreBrand, updateBrand } from "../controllers/brandController.js";
 const brandRouter = express.Router();
 
 brandRouter.get("/", getBrands);
 
-brandRouter.post("/", passport.authenticate("jwt", { session: false }), checkAdmin, createBrand);
-brandRouter.get("/trash", passport.authenticate("jwt", { session: false }), checkAdmin, getBrandsFromTrash);
+brandRouter.post("/", requireAuth, checkAdmin, createBrand);
+brandRouter.get("/trash", requireAuth, checkAdmin, getBrandsFromTrash);
 
 brandRouter.get("/:slug", getBrandBySlug);
-brandRouter.put("/:slug", passport.authenticate("jwt", { session: false }), checkAdmin, updateBrand);
-brandRouter.delete("/:slug", passport.authenticate("jwt", { session: false }), checkAdmin, deleteBrand);
-brandRouter.patch("/:slug/restore", passport.authenticate("jwt", { session: false }), checkAdmin, restoreBrand);
+brandRouter.put("/:slug", requireAuth, checkAdmin, updateBrand);
+brandRouter.delete("/:slug", requireAuth, checkAdmin, deleteBrand);
+brandRouter.patch("/:slug/restore", requireAuth, checkAdmin, restoreBrand);
 
 export default brandRouter;
diff --git a/apps/api/routes/cartRoutes.js b/apps/api/routes/cartRoutes.js
index f066cb0..b8b2d40 100644
--- a/apps/api/routes/cartRoutes.js
+++ b/apps/api/routes/cartRoutes.js
@@ -1,9 +1,10 @@
 import express from "express";
 import { addToCart, updateCart, removeFromCart, getCartSummary } from "../controllers/cartController.js";
 import { passport } from "../utils/passport.js";
+import { requireAuth } from "../middleware/authMiddleware.js";
 
 const cartRoutes = express.Router();
-cartRoutes.use(passport.authenticate("jwt", { session: false }));
+cartRoutes.use(requireAuth);
 
 cartRoutes.post("/", addToCart);
 cartRoutes.put("/", updateCart);
diff --git a/apps/api/routes/categoryRoutes.js b/apps/api/routes/categoryRoutes.js
index 77c93aa..986704b 100644
--- a/apps/api/routes/categoryRoutes.js
+++ b/apps/api/routes/categoryRoutes.js
@@ -1,52 +1,19 @@
-import express from 'express';
-import checkAdmin from '../middleware/adminMiddleware.js';
-import {
-  createCategory,
-  deleteCategory,
-  getCategories,
-  getCategoriesFromTrash,
-  getCategoryBySlug,
-  restoreCategory,
-  updateCategory,
-} from '../controllers/categoryController.js';
-import { passport } from '../utils/passport.js';
-import { validate } from '../middleware/zodMiddleware.js';
-import { categorySchema } from '../schemas/categorySchema.js';
+import express from "express";
+import checkAdmin from "../middleware/adminMiddleware.js";
+import { requireAuth } from "../middleware/authMiddleware.js";
+import { createCategory, deleteCategory, getCategories, getCategoriesFromTrash, getCategoryBySlug, restoreCategory, updateCategory } from "../controllers/categoryController.js";
+import { passport } from "../utils/passport.js";
+import { validate } from "../middleware/zodMiddleware.js";
+import { categorySchema } from "../schemas/categorySchema.js";
 
 const categoryRouter = express.Router();
 
-categoryRouter.post(
-  '/',
-  passport.authenticate('jwt', { session: false }),
-  checkAdmin,
-  validate(categorySchema),
-  createCategory
-);
-categoryRouter.get('/', getCategories);
-categoryRouter.get(
-  '/trash',
-  passport.authenticate('jwt', { session: false }),
-  checkAdmin,
-  getCategoriesFromTrash
-);
-categoryRouter.get('/:slug', getCategoryBySlug);
-categoryRouter.put(
-  '/:slug',
-  passport.authenticate('jwt', { session: false }),
-  checkAdmin,
-  updateCategory
-);
-categoryRouter.delete(
-  '/:slug',
-  passport.authenticate('jwt', { session: false }),
-  checkAdmin,
-  deleteCategory
-);
-categoryRouter.patch(
-  '/:slug/restore',
-  passport.authenticate('jwt', { session: false }),
-  checkAdmin,
-  restoreCategory
-);
+categoryRouter.post("/", requireAuth, checkAdmin, validate(categorySchema), createCategory);
+categoryRouter.get("/", getCategories);
+categoryRouter.get("/trash", requireAuth, checkAdmin, getCategoriesFromTrash);
+categoryRouter.get("/:slug", getCategoryBySlug);
+categoryRouter.put("/:slug", requireAuth, checkAdmin, updateCategory);
+categoryRouter.delete("/:slug", requireAuth, checkAdmin, deleteCategory);
+categoryRouter.patch("/:slug/restore", requireAuth, checkAdmin, restoreCategory);
 
 export default categoryRouter;
diff --git a/apps/api/routes/router.js b/apps/api/routes/router.js
index c4a0d8a..c97b415 100644
--- a/apps/api/routes/router.js
+++ b/apps/api/routes/router.js
@@ -1,31 +1,32 @@
-import express from 'express';
+import express from "express";
 const router = express.Router();
-import authRouter from './authRoutes.js';
-import userRouter from './userRoutes.js';
-import adminRouter from './adminRoutes.js';
-import cartRoutes from './cartRoutes.js';
-import orderRoutes from './orderRoutes.js';
-import discountRoutes from './discountRoutes.js';
-import brandRouter from './brandRoutes.js';
-import categoryRouter from './categoryRoutes.js';
-import { passport } from '../utils/passport.js';
-import checkAdmin from '../middleware/adminMiddleware.js';
-import shippingInfoRouter from './shippingInfoRoutes.js';
+import authRouter from "./authRoutes.js";
+import userRouter from "./userRoutes.js";
+import adminRouter from "./adminRoutes.js";
+import cartRoutes from "./cartRoutes.js";
+import orderRoutes from "./orderRoutes.js";
+import discountRoutes from "./discountRoutes.js";
+import brandRouter from "./brandRoutes.js";
+import categoryRouter from "./categoryRoutes.js";
+import { passport } from "../utils/passport.js";
+import checkAdmin from "../middleware/adminMiddleware.js";
+import { requireAuth } from "../middleware/authMiddleware.js";
+import shippingInfoRouter from "./shippingInfoRoutes.js";
 
 // router -> controller -> service
 
-router.get('/', (req, res) => {
-  res.send('Hello, World!');
+router.get("/", (req, res) => {
+  res.send("Hello, World!");
 });
 
-router.use('/admin', passport.authenticate('jwt', { session: false }), checkAdmin, adminRouter);
-router.use('/brand', brandRouter);
-router.use('/category', categoryRouter);
-router.use('/auth', authRouter);
-router.use('/user', userRouter);
-router.use('/cart', cartRoutes);
-router.use('/order', orderRoutes);
-router.use('/discount', discountRoutes);
-router.use('/shipping', shippingInfoRouter);
+router.use("/admin", requireAuth, checkAdmin, adminRouter);
+router.use("/brand", brandRouter);
+router.use("/category", categoryRouter);
+router.use("/auth", authRouter);
+router.use("/user", userRouter);
+router.use("/cart", cartRoutes);
+router.use("/order", orderRoutes);
+router.use("/discount", discountRoutes);
+router.use("/shipping", shippingInfoRouter);
 
 export default router;
diff --git a/apps/api/routes/shippingInfoRoutes.js b/apps/api/routes/shippingInfoRoutes.js
index e9498fc..1cd572e 100644
--- a/apps/api/routes/shippingInfoRoutes.js
+++ b/apps/api/routes/shippingInfoRoutes.js
@@ -1,11 +1,11 @@
 import express from "express";
-import { changeIsDefault, createShippingAddress, deleteShippingInfo, getAllShippingInfo } from "../controllers/shippingInfoController.js";
-import { passport } from "../utils/passport.js";
 import { validate } from "../middleware/zodMiddleware.js";
+import { requireAuth } from "../middleware/authMiddleware.js";
 import { ShippingInfoSchema } from "../schemas/shippingInfoSchema.js";
+import { changeIsDefault, createShippingAddress, deleteShippingInfo, getAllShippingInfo } from "../controllers/shippingInfoController.js";
 
 const shippingInfoRouter = express.Router();
-shippingInfoRouter.use(passport.authenticate("jwt", { session: false }));
+shippingInfoRouter.use(requireAuth);
 
 shippingInfoRouter.get("/", getAllShippingInfo);
 shippingInfoRouter.post("/", validate(ShippingInfoSchema), createShippingAddress);
diff --git a/apps/api/services/authService.js b/apps/api/services/authService.js
index 8b10e5f..5016990 100644
--- a/apps/api/services/authService.js
+++ b/apps/api/services/authService.js
@@ -1,10 +1,18 @@
-import bcrypt from 'bcryptjs';
-import prisma from '../prisma/prismaClient.js';
-import jwt from 'jsonwebtoken';
+import bcrypt from "bcryptjs";
+import prisma from "../prisma/prismaClient.js";
+import jwt from "jsonwebtoken";
+import { nanoid } from "nanoid";
+
+export const signTokensService = ({ id, role }) => {
+  return {
+    refreshToken: jwt.sign({ sub: id }, process.env.JWT_SECRET, { expiresIn: "15m" }),
+    accessToken: jwt.sign({ jti: nanoid(), sub: id, role }, process.env.JWT_REFRESH_SECRET, { expiresIn: "7d" }),
+  };
+};
 
 export const registerService = async (email, password, fullName) => {
   if (await prisma.user.findUnique({ where: { email } })) {
-    throw new Error('Email already exists');
+    throw new Error("Email already exists");
   }
   const hashedPassword = await bcrypt.hash(password, 10);
   return await prisma.user.create({
@@ -12,16 +20,21 @@ export const registerService = async (email, password, fullName) => {
   });
 };
 
-export const refreshService = async (refreshToken) => {
-  if (!refreshToken) {
-    throw new Error('No token provided');
+export const refreshService = async (oldRefreshToken) => {
+  if (!oldRefreshToken) {
+    throw new Error("No token provided");
   }
 
-  const { sub } = jwt.verify(refreshToken, process.env.JWT_REFRESH_SECRET);
+  const { sub } = jwt.verify(oldRefreshToken, process.env.JWT_REFRESH_SECRET);
   const user = await prisma.user.findUnique({ where: { id: sub } });
   if (!user) {
-    throw new Error('Invalid token');
+    throw new Error("Invalid token");
   }
 
-  return user;
+  return signTokensService(user);
+};
+
+export const logoutService = async (refreshToken) => {
+  const { jti } = jwt.verify(refreshToken, process.env.JWT_REFRESH_SECRET);
+  await prisma.revokedToken.create({ data: { jti } });
 };
diff --git a/apps/api/utils/passport.js b/apps/api/utils/passport.js
index 5812657..729e10b 100644
--- a/apps/api/utils/passport.js
+++ b/apps/api/utils/passport.js
@@ -1,31 +1,10 @@
-import passport from 'passport';
-import jwt from 'jsonwebtoken';
-import prisma from '../prisma/prismaClient.js';
-import bcrypt from 'bcryptjs';
-import { Strategy as LocalStrategy } from 'passport-local';
-import { Strategy as GoogleStrategy } from 'passport-google-oauth20';
-import { Strategy as FacebookStrategy } from 'passport-facebook';
-import { Strategy as JwtStrategy, ExtractJwt } from 'passport-jwt';
-
-function issueTokens(res, sub) {
-  const refreshToken = jwt.sign({ sub }, process.env.JWT_REFRESH_SECRET, {
-    expiresIn: '7d',
-  });
-  const accessToken = jwt.sign({ sub }, process.env.JWT_SECRET, {
-    expiresIn: '15m',
-  });
-  console.log("SUB: ", sub)
-
-  res.cookie('refreshToken', refreshToken, {
-    httpOnly: true,
-    secure: process.env.NODE_ENV === 'production',
-    sameSite: 'strict',
-    path: '/auth/refresh',
-    maxAge: 7 * 24 * 60 * 60 * 1000,
-  });
-
-  return res.json({ accessToken });
-}
+import passport from "passport";
+import prisma from "../prisma/prismaClient.js";
+import bcrypt from "bcryptjs";
+import { Strategy as LocalStrategy } from "passport-local";
+import { Strategy as GoogleStrategy } from "passport-google-oauth20";
+import { Strategy as FacebookStrategy } from "passport-facebook";
+import { Strategy as JwtStrategy, ExtractJwt } from "passport-jwt";
 
 passport.use(
   new JwtStrategy(
@@ -37,35 +16,28 @@ passport.use(
       try {
         const user = await prisma.user.findUnique({ where: { id: sub } });
         if (!user) {
-          return done(null, false, { message: 'No token provided' });
+          return done(null, false, { message: "No token provided" });
         }
         return done(null, user);
       } catch (err) {
         return done(err);
       }
-    }
-  )
+    },
+  ),
 );
 
 passport.use(
-  new LocalStrategy(
-    { usernameField: 'email', passwordField: 'password' },
-    async (email, password, done) => {
-      try {
-        const user = await prisma.user.findUnique({ where: { email } });
-        if (
-          !user ||
-          !user.password ||
-          !(await bcrypt.compare(password, user.password))
-        ) {
-          return done(null, false, { message: 'Incorrect email or password' });
-        }
-        return done(null, user);
-      } catch (err) {
-        return done(err);
+  new LocalStrategy({ usernameField: "email", passwordField: "password" }, async (email, password, done) => {
+    try {
+      const user = await prisma.user.findUnique({ where: { email } });
+      if (!user || !user.password || !(await bcrypt.compare(password, user.password))) {
+        return done(null, false, { message: "Incorrect email or password" });
       }
+      return done(null, user);
+    } catch (err) {
+      return done(err);
     }
-  )
+  }),
 );
 
 passport.use(
@@ -73,36 +45,31 @@ passport.use(
     {
       clientID: process.env.GOOGLE_CLIENT_ID,
       clientSecret: process.env.GOOGLE_CLIENT_SECRET,
-      callbackURL: '/auth/google/callback',
+      callbackURL: "/auth/google/callback",
     },
-    async (
-      accessToken,
-      refreshToken,
-      { id: providerId, displayName: fullName, emails },
-      done
-    ) => {
-      const email = emails?.[0]?.value.toLowerCase() ?? '';
+    async (accessToken, refreshToken, { id: providerId, displayName: fullName, emails }, done) => {
+      const email = emails?.[0]?.value.toLowerCase() ?? "";
 
       try {
         let user = await prisma.user.findUnique({ where: { email } });
         if (!user) {
           user = await prisma.user.create({
-            data: { email, fullName, role: 'CUSTOMER' },
+            data: { email, fullName, role: "CUSTOMER" },
           });
         }
 
         await prisma.oAuthAccount.upsert({
-          where: { provider_providerId: { provider: 'google', providerId } },
+          where: { provider_providerId: { provider: "google", providerId } },
           update: {},
-          create: { provider: 'google', providerId, userId: user.id },
+          create: { provider: "google", providerId, userId: user.id },
         });
 
         done(null, user);
       } catch (error) {
         return done(error);
       }
-    }
-  )
+    },
+  ),
 );
 
 passport.use(
@@ -110,38 +77,33 @@ passport.use(
     {
       clientID: process.env.FACEBOOK_APP_ID,
       clientSecret: process.env.FACEBOOK_APP_SECRET,
-      callbackURL: '/auth/facebook/callback',
-      profileFields: ['id', 'emails', 'name'],
+      callbackURL: "/auth/facebook/callback",
+      profileFields: ["id", "emails", "name"],
     },
-    async (
-      accessToken,
-      refreshToken,
-      { id: providerId, name: { givenName, familyName }, emails },
-      done
-    ) => {
-      const email = emails?.[0]?.value.toLowerCase() ?? '';
-      const fullName = [givenName, familyName].join(' ');
+    async (accessToken, refreshToken, { id: providerId, name: { givenName, familyName }, emails }, done) => {
+      const email = emails?.[0]?.value.toLowerCase() ?? "";
+      const fullName = [givenName, familyName].join(" ");
 
       try {
         let user = await prisma.user.findUnique({ where: { email } });
         if (!user) {
           user = await prisma.user.create({
-            data: { email, fullName, role: 'CUSTOMER' },
+            data: { email, fullName, role: "CUSTOMER" },
           });
         }
 
         await prisma.oAuthAccount.upsert({
-          where: { provider_providerId: { provider: 'facebook', providerId } },
+          where: { provider_providerId: { provider: "facebook", providerId } },
           update: {},
-          create: { provider: 'facebook', providerId, userId: user.id },
+          create: { provider: "facebook", providerId, userId: user.id },
         });
 
         done(null, user);
       } catch (err) {
         done(err);
       }
-    }
-  )
+    },
+  ),
 );
 
 passport.serializeUser((user, done) => {
@@ -157,4 +119,4 @@ passport.deserializeUser(async (id, done) => {
   }
 });
 
-export { passport, issueTokens };
+export { passport };
diff --git a/package.json b/package.json
index e406c29..ab85fc7 100644
--- a/package.json
+++ b/package.json
@@ -14,14 +14,5 @@
   "workspaces": [
     "apps/*",
     "packages/*"
-  ],
-  "dependencies": {
-    "bcryptjs": "^3.0.2",
-    "cors": "^2.8.5",
-    "express": "^5.1.0",
-    "jsonwebtoken": "^9.0.2",
-    "node-cron": "^4.2.1",
-    "nodemailer": "^7.0.6",
-    "nodemon": "^3.1.10"
-  }
-}
+  ]
+}
\ No newline at end of file
